onedal/svm/svm.cpp

```diff
@@ -0,0 +1,227 @@
+            ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
+        }
+    }
+
+    Ops ops;
+};
+
+template <typename Ops>
+struct kernel2t {
+    kernel2t(const Ops& ops)
+        : ops(ops) {}
+
+    template <typename Float, typename Method>
+    auto operator()(const py::dict& params) {
+        using namespace svm;
+
+        auto kernel = params["kernel"].cast<std::string>();
+        ONEDAL_PARAM_DISPATCH_VALUE(kernel, "linear", ops, Float, Method, linear_kernel::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(kernel, "rbf", ops, Float, Method, rbf_kernel::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(kernel, "poly", ops, Float, Method, polynomial_kernel::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_VALUE(kernel, "sigmoid", ops, Float, Method, sigmoid_kernel::descriptor<Float>);
+        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(kernel);
+    }
+
+    Ops ops;
+};
+
+struct params2desc {
+    template <typename Float, typename Method, typename Task, typename Kernel>
+    auto operator()(const pybind11::dict& params) {
+        using namespace svm;
+
+        constexpr bool is_cls = std::is_same_v<Task, task::classification>;
+        constexpr bool is_nu_cls = std::is_same_v<Task, task::nu_classification>;
+        constexpr bool is_reg = std::is_same_v<Task, task::regression>;
+        constexpr bool is_nu_reg = std::is_same_v<Task, task::nu_regression>;
+
+        auto desc = descriptor<Float, Method, Task, Kernel>{ get_kernel_descriptor<Kernel>(params) }
+                        .set_max_iteration_count(params["max_iteration_count"].cast<std::int64_t>())
+                        .set_accuracy_threshold(params["accuracy_threshold"].cast<double>())
+                        .set_cache_size(params["cache_size"].cast<double>())
+                        .set_tau(params["tau"].cast<double>())
+                        .set_shrinking(params["shrinking"].cast<bool>());
+
+        if constexpr (is_cls || is_reg || is_nu_reg) {
+            desc.set_c(params["c"].cast<double>());
+        }
+        if constexpr (is_cls || is_nu_cls) {
+            desc.set_class_count(params["class_count"].cast<std::int64_t>());
+        }
+        if constexpr (is_reg) {
+            desc.set_epsilon(params["epsilon"].cast<double>());
+        }
+        if constexpr (is_nu_reg || is_nu_cls) {
+            desc.set_nu(params["nu"].cast<double>());
+        }
+
+        return desc;
+    }
+};
+
+template <typename Policy, typename Task>
+void init_train_ops(py::module_& m) {
+    m.def("train",
+          [](const Policy& policy,
+             const py::dict& params,
+             const table& data,
+             const table& responses,
+             const table& weights) {
+              using namespace svm;
+              using input_t = train_input<Task>;
+
+              train_ops ops(policy, input_t{ data, responses, weights }, params2desc{} );
```

Comment:
I would recommend explicitly writing:
```python
svm::train_ops(...)
```

Reply:
`train_ops` is not algorithm-specific: this is a single class for the whole library.

Reply:
ok, svm namespace need for `train_input`only? maybe without svm namespace?